### PR TITLE
[AMDGPU][NewPM] Port SIShrinkInstructions to new pass manager.

### PR DESCRIPTION
**Pull request by cdevadas**

### CODE REVIEW

**Comment by cdevadas** [comment]
* **#106967** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/106967?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @cdevadas and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: llvm/lib/Target/AMDGPU/SIShrinkInstructions.cpp [hunk 1]
```diff
@@ -1077,3 +1081,21 @@ bool SIShrinkInstructions::runOnMachineFunction(MachineFunction &MF) {
...
+  if (skipFunction(MF.getFunction()))
+    return false;
+
+  return SIShrinkInstructions().run(MF);
+}
+
+PreservedAnalyses
+SIShrinkInstructionsPass::run(MachineFunction &MF,
+                              MachineFunctionAnalysisManager &) {
+  if (MF.getFunction().hasOptNone() || !SIShrinkInstructions().run(MF))
```
Comment (pravinjagtap) [inline llvm/lib/Target/AMDGPU/SIShrinkInstructions.cpp:1095]: Shouldn't codegen pipeline builder take care of opt none? I am not sure when to use TM and Fcn attribute to decide pass invocatoin.

**Review by pravinjagtap — COMMENTED** [review]
LGTM

**Review by pravinjagtap — APPROVED** [review]

File: llvm/lib/Target/AMDGPU/SIShrinkInstructions.cpp (hunk 1 above)
Comment (cdevadas) [reply llvm/lib/Target/AMDGPU/SIShrinkInstructions.cpp:1095]: This early exit is mainly to skip the optimization passes for -O0. This specific check hasOptNone checks for the function attribute Attribute::OptimizeNone, set to true for -O0 opt.levels. The attributes and metadata are still in the LLVM IR and not available in the MIR.

**Comment by cdevadas** [comment]
### Merge activity

* **Sep 3, 1:21 AM EDT**: @cdevadas started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/106967).

**Pull request merged**